File: 0433_Minimum Genetic Mutation/MinimumGeneticMutation_DFS.cs
public class Solution {
    
    private int ans = int.MaxValue;
    private char[] GENES = new char[]{'A','C','G','T'};
    
    public int MinMutation(string start, string end, string[] bank) {
        if(bank == null || bank.Length == 0) return -1;
        
        var graph = new HashSet<string>();
        foreach(var b in bank)
        {
            graph.Add(b);
        }
        graph.Add(start);
        
        var visited = new HashSet<string>();
        
        DFS(start, end, graph, visited, 0);
        
        return ans == int.MaxValue ? -1 : ans;
    }
    
    private void DFS(string src, string end, HashSet<string> graph, HashSet<string> visited, int mutations)
    {
        
        if(!graph.Contains(src) || visited.Contains(src)) return;
        
        if(src == end)
        {
            ans = Math.Min(mutations, ans);
            return;
        }
        
        
        visited.Add(src);
        
        var char_arr = src.ToCharArray();
        
        for(int i=0;i<src.Length;i++)
        {
            for(int j=0;j<GENES.Length;j++)
            {
                var rawChar = char_arr[i];
                char_arr[i] = GENES[j];
                var newStr = new String(char_arr);
                DFS(newStr, end, graph, visited, mutations+1);
                char_arr[i] = rawChar;
            }
        }
    }
}

File: 0433_Minimum Genetic Mutation/MinimumGeneticMutation_DFS.java
class Solution {
    private int ans = Integer.MAX_VALUE;
    
    public int minMutation(String start, String end, String[] bank) {
        if(start.equals(end)) return 0;
        char[] GENES = new char[]{'A','C','G','T'};
        Set<String> banks = new HashSet<String>();
        for(String b : bank)  banks.add(b);
        if(!banks.contains(end)) return -1;
        banks.add(start);
        
        DFS(banks, new HashSet(), start, 0, GENES, end);
        
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }
    
    private void DFS(Set<String> banks, Set<String> visited, String g, int mutations, char[] GENES, String end) {
        if(!banks.contains(g) || visited.contains(g)) return;
        if(g.equals(end)) {
            ans = Math.min(mutations, ans);
            return;
        }
        
        visited.add(g);
        
        char[] chars = g.toCharArray();
        
        for(int i=0;i<chars.length;i++) {
            char oldChar = chars[i];
            for(int j=0; j < GENES.length; j++) {
                if(oldChar == GENES[j]) continue;
                chars[i] = GENES[j];
                String newGene = new String(chars);
                DFS(banks, visited, newGene, mutations+1, GENES, end);
            }
            chars[i] = oldChar;
        }
    }
}